api/types/maintenancewindow.go

```diff
@@ -0,0 +1,222 @@
+	}
+}
+
+// Export exports the next `n` upgrade windows as a schedule object, starting from `from`.
+func (w *AgentUpgradeWindow) Export(from time.Time, n int) AgentUpgradeSchedule {
+	gen := w.generator(from)
+
+	sched := AgentUpgradeSchedule{
+		Windows: make([]ScheduledAgentUpgradeWindow, 0, n),
+	}
+	for i := 0; i < n; i++ {
+		start, stop := gen()
+		sched.Windows = append(sched.Windows, ScheduledAgentUpgradeWindow{
+			Start: start.UTC(),
+			Stop:  stop.UTC(),
+		})
+	}
+
+	return sched
+}
+
+func (s *AgentUpgradeSchedule) Clone() *AgentUpgradeSchedule {
+	return proto.Clone(s).(*AgentUpgradeSchedule)
+}
+
+// NewMaintenanceWindow creates a new maintenance window with no parameters set.
+func NewMaintenanceWindow() MaintenanceWindow {
+	var mw MaintenanceWindowV1
+	mw.setStaticFields()
+	return &mw
+}
+
+type MaintenanceWindow interface {
+	Resource
+
+	// GetNonce gets the nonce of the maintenance window.
+	GetNonce() uint64
+
+	// WithNonce creates a shallow copy with a new nonce.
+	WithNonce(nonce uint64) any
+
+	// GetAgentUpgradeWindow gets the agent upgrade window.
+	GetAgentUpgradeWindow() (win AgentUpgradeWindow, ok bool)
+
+	// SetAgentUpgradeWindow sets the agent upgrade window.
+	SetAgentUpgradeWindow(win AgentUpgradeWindow)
+
+	CheckAndSetDefaults() error
+}
+
+func (m *MaintenanceWindowV1) setStaticFields() {
+	if m.Version == "" {
+		m.Version = V1
+	}
+
+	if m.Kind == "" {
+		m.Kind = KindMaintenanceWindow
+	}
+
+	if m.Metadata.Name == "" {
+		m.Metadata.Name = MetaNameMaintenanceWindow
+	}
+}
+
+func (m *MaintenanceWindowV1) CheckAndSetDefaults() error {
+	m.setStaticFields()
+
+	if err := m.ResourceHeader.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if m.Version != V1 {
+		return trace.BadParameter("unexpected maintenance window resource version %q (expected %q)", m.Version, V1)
+	}
+
+	if m.Kind == MetaNameMaintenanceWindow {
+		// normalize easy mixup
+		m.Kind = KindMaintenanceWindow
+	}
+
+	if m.Kind != KindMaintenanceWindow {
+		return trace.BadParameter("unexpected maintenance window kind %q (expected %q)", m.Kind, KindMaintenanceWindow)
+	}
+
+	if m.Metadata.Name == KindMaintenanceWindow {
+		// normalize easy mixup
+		m.Metadata.Name = MetaNameMaintenanceWindow
+	}
+
+	if m.Metadata.Name != MetaNameMaintenanceWindow {
+		return trace.BadParameter("unexpected maintenance window name %q (expected %q)", m.Metadata.Name, MetaNameMaintenanceWindow)
+	}
+
+	if m.Spec.AgentUpgrades != nil {
+		if h := m.Spec.AgentUpgrades.UTCStartHour; h > 23 {
+			return trace.BadParameter("agent upgrade window utc start hour must be in range 0..23, got %d", h)
+		}
+
+		for _, day := range m.Spec.AgentUpgrades.Weekdays {
+			if _, ok := parseWeekday(day); !ok {
+				return trace.BadParameter("invalid weekday in agent upgrade window: %q", day)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *MaintenanceWindowV1) GetNonce() uint64 {
+	return m.Nonce
+}
+
+func (m *MaintenanceWindowV1) WithNonce(nonce uint64) any {
```

Comment:
Why do you erase the type here?

Reply:
I did this because I wanted to include `WithNonce` in the `MaintenanceWindow` interface.  We could alternatively implement the `NonceProtectedResource` interface with a generic parameter.  Ex:

```golang
type NonceProtectedResource[T any] interface {
	GetNonce() uint64
	WithNonce(uint64) T
	// ...other methods
}


// example usage
func Increment[T NonceProtectedResource[T]](rsc T) T {
	return rsc.WithNonce(rsc.GetNonce() + 1)
}

```

This would eliminate the need to do type-casting inside functions that are generic over the `NonceProtectedResource` interface, but would hide the fact that `MaintenanceWindow` types are expected to implement `WithNonce`.

IDK what the right answer is here really. Typically when I see `any` in a signature, I take it as a warning that I need to be careful about my concrete types, because some type-casting is going to happen somewhere.  Because of that, I opted to take the `WithNonce(uint64) any` route... just because it at least forces an implementer of the `MaintenanceWindow` interface to _confront_ the fact that their type is going to be expected to play nicely in a context where type-casting is happening.

Reply:
I feel like we are overcomplicating things here. Do we really need generic interfaces to build an API for setting a maintenance window? If we were writing this code 18 months ago and generics were not available what would we have done instead?

Reply:
> Do we really need generic interfaces to build an API for setting a maintenance window?

No, but we've started building up some useful tooling in `local/generic` to make working with resource CRUD a bit simpler.  I don't see why we wouldn't want to make useful patterns available there for future reuse if we have the option.

> If we were writing this code 18 months ago and generics were not available what would we have done instead?

Not much different... replace `T` with `*types.MaintenanceWindowV1` and get rid of a three-method interface.  The shallow copying method would _still_ be doing type-erasure... it would return its parent interface (`types.MaintenanceWindow`) instead of `any`, just like the `Clone()` methods currently do, but that would still mean us returning an interface type, and would still mean we'd need to downcast to a concrete type if we wanted to be sure the type we are writing into the backend is the same as the type we are expecting to read.